chains/evm/contracts/pools/USDC/USDCTokenPoolProxy.sol

```diff
@@ -0,0 +1,372 @@
+        // destinationCaller (the message transmitter proxy), we can route the message to the v1 pool, but we first
+        // need to turn the source pool data into the new format, otherwise the decoding scheme will fail. Once there is
+        // confidence that no more messages are inflight, these branches can be safely removed.
+
+        // Since the CCTP v1 pool will have this contract set as an allowed caller, no additional configurations are
+        // needed to route the message to the v1 pool.
+        return USDCTokenPool(s_pools.cctpV1Pool).releaseOrMint(_generateNewReleaseOrMintIn(releaseOrMintIn));
+      }
+    }
+
+    // The first 4 bytes of source pool data are the version which can be extracted directly and cast into a uint32.
+    bytes4 version = bytes4(releaseOrMintIn.sourcePoolData[:4]);
+
+    // If the source pool data is the lock release flag, use the lock release pool set for the remote chain selector.
+    if (version == USDCSourcePoolDataCodec.LOCK_RELEASE_FLAG) {
+      return USDCTokenPool(s_lockReleasePools[releaseOrMintIn.remoteChainSelector]).releaseOrMint(releaseOrMintIn);
+    }
+
+    if (version == USDCSourcePoolDataCodec.CCTP_VERSION_1_TAG) {
+      return USDCTokenPool(s_pools.cctpV1Pool).releaseOrMint(releaseOrMintIn);
+    }
+
+    // Both tags will route to the same CCTP V2 pool, but will allow for pools to have greater granularity in deciding
+    // the type of transfer (slow or fast) to use when depositing into CCTP.
+    if (
+      version == USDCSourcePoolDataCodec.CCTP_VERSION_2_TAG
+        || version == USDCSourcePoolDataCodec.CCTP_VERSION_2_FAST_TRANSFER_TAG
+    ) {
+      return USDCTokenPool(s_pools.cctpV2Pool).releaseOrMint(releaseOrMintIn);
+    }
+
+    revert InvalidMessageVersion(version);
+  }
+
+  /// @notice Update the pool addresses that this token pool will route a message to.
+  /// @param pools The new pool addresses to update the token pool proxy with. Since the legacy CCTP V1 pool may not be
+  /// used, the zero address is a valid input and therefore input sanitization for it is not required.
+  function updatePoolAddresses(
+    PoolAddresses calldata pools
+  ) external onlyOwner {
+    if (pools.cctpV1Pool == address(0) || pools.cctpV2Pool == address(0)) {
+      revert AddressCannotBeZero();
+    }
+
+    // If the V1 or V2 Pool does not support the IPoolV1 interface, revert.
+    // If the legacy CCTP V1 Pool is being used, then it must support the IPoolV1 interface. If it is not, don't check it.
+    if (
+      !pools.cctpV1Pool._supportsInterfaceReverting(type(IPoolV1).interfaceId)
+        || !pools.cctpV2Pool._supportsInterfaceReverting(type(IPoolV1).interfaceId)
+        || (
+          pools.legacyCctpV1Pool != address(0)
+            && !pools.legacyCctpV1Pool._supportsInterfaceReverting(type(IPoolV1).interfaceId)
+        )
+    ) {
+      revert TokenPoolUnsupported();
```

Comment:
Can we split these checks and have the error return the failing one?

Reply:
https://github.com/smartcontractkit/chainlink-ccip/pull/1134/commits/4aa8aa0fd730a2c6fa961a4f1bc5ad35a65fb086#diff-7d40f3f52c8e578caab476a14e440188e432d8f912bfa3f6871375cc86ac5e4eR223